### PR TITLE
[WIP] adds BucketSampler

### LINKED ISSUE #1474: Add bucket sampler
## 🚀 Feature

**Motivation**

The legacy `BucketIterator` was convenient because it could batch samples by length to minimize padding. It had many disadvantages because of its API and non-comformance with other parts of the pytorch data{sets,loader} ecosystem. It would be nice if torchtext supported the spirit of the `BucketIterator` by way of a `Sampler`.

**Pitch**

A sampler with the ability to specify maximum bucket size should be added similar to those in torchnlp and allennlp. This can be used with existing datasets in torchtext but as a kwarg to the pytorch `DataLoader` so sampling minimizes padding.

**Alternatives**

Users who want this functionality need to implement their own samplers.

**Additional context**

N/A

### PR DESCRIPTION
**Pull request by erip**
Fixes #1474 

TODO: 
- [ ] tests
- [ ] benchmark

### CODE REVIEW

**Comment by erip** [comment]
It seems like pytorch dropped support for 3.6 in nightly.

**Comment by erip** [comment]
I'm going to close this since it'll be OBE by torchdata's bucketize datapipes.
